File: topos/graphs/uast/mapper_cpp.py
from __future__ import annotations

from tree_sitter import Node

from topos.graphs.uast.mapper_common import map_tree_sitter_to_uast
from topos.graphs.uast.models import UASTNode

# Real tree-sitter-cpp grammar node names (verified against the vendored
# grammar directly — the previous dict here was copy-pasted from the
# Python/Rust mappers and used node names tree-sitter-cpp doesn't emit at
# all, e.g. `class_definition`/`struct_item`, so every C++ type
# declaration silently fell to `Unknown`; see issue #158).
_DECLARATION_TYPES = {
    "function_definition": "FunctionDecl",  # free functions AND in-class
    # method *definitions* (with a body) — tree-sitter-cpp doesn't
    # distinguish the two at this node.
    "class_specifier": "TypeDecl",
    "struct_specifier": "TypeDecl",
    "enum_specifier": "TypeDecl",
    "union_specifier": "TypeDecl",
}

# `declaration` (free-standing) and `field_declaration` (inside a class
# body) are dual-purpose in the grammar: a variable/member declaration
# (`int x;`) and a declaration-only function/method signature with no
# body (`void f(int);`, or a pure-virtual `virtual double area() const =
# 0;`) both use the same wrapper node, distinguished only by whether a
# `function_declarator` child is present.
_DECLARATION_ONLY_KINDS = {
    "declaration": "VarDecl",
    "field_declaration": "VarDecl",
}


def _has_function_declarator(node: Node) -> bool:
    return any(child.type == "function_declarator" for child in node.children)


_STATEMENT_TYPES = {
    "if_statement": "IfStmt",
    "for_statement": "ForStmt",
    "while_statement": "WhileStmt",
    "match_statement": "MatchStmt",
    "return_statement": "ReturnStmt",
    "break_statement": "BreakStmt",
    "continue_statement": "ContinueStmt",
    "throw_statement": "ThrowStmt",
    "try_statement": "TryStmt",
    "expression_statement": "ExprStmt",
}

_EXPRESSION_TYPES = {
    "assignment_expression": "AssignExpr",
    "binary_expression": "BinaryExpr",
    "unary_expression": "UnaryExpr",
    "call_expression": "CallExpr",
    "field_expression": "MemberExpr",
    "subscript_expression": "MemberExpr",
}


def map_node_kind(node: Node) -> str:
    if node.type in _DECLARATION_TYPES:
        return _DECLARATION_TYPES[node.type]
    if node.type in _DECLARATION_ONLY_KINDS:
        return "FunctionDecl" if _has_function_declarator(node) else "VarDecl"
    if node.type in _STATEMENT_TYPES:
        return _STATEMENT_TYPES[node.type]
    if node.type in _EXPRESSION_TYPES:
        return _EXPRESSION_TYPES[node.type]
    if node.type == "identifier":
        return "Identifier"
    if node.type.endswith("literal") or node.type in {"string", "integer", "float"}:
        return "Literal"
    if node.type in {"module", "program", "translation_unit", "source_file"}:
        return "File"
    return "Unknown"


# Martin Abstractness classification (issue #124/#158): a class/struct is
# abstract iff it declares at least one pure-virtual method
# (`virtual ... = 0;`) — the C++ idiom for an interface/abstract base
# class. `enum`/`union` are always concrete.
_TYPE_KIND = {
    "class_specifier": "class",
    "struct_specifier": "struct",
    "enum_specifier": "enum",
    "union_specifier": "union",
}


def _is_pure_virtual(field_decl: Node) -> bool:
    """True for a declaration-only method signature with a `= 0`
    pure-specifier, e.g. `virtual double area() const = 0;`."""
    if not _has_function_declarator(field_decl):
        return False
    return any(
        child.type == "number_literal" and child.text == b"0"
        for child in field_decl.named_children
    )


def _has_pure_virtual_method(type_node: Node) -> bool:
    for child in type_node.children:
        if child.type != "field_declaration_list":
            continue
        return any(
            member.type == "field_declaration" and _is_pure_virtual(member)
            for member in child.named_children
        )
    return False


def extract_type_attributes(node: Node) -> dict[str, object]:
    type_kind = _TYPE_KIND.get(node.type)
    if type_kind is None:
        return {}
    is_class_like = node.type in {"class_specifier", "struct_specifier"}
    if is_class_like and _has_pure_virtual_method(node):
        return {"typeKind": "abstractClass"}
    return {"typeKind": type_kind}


def map_cpp_tree_to_uast(root: Node, file: str | None = None) -> UASTNode:
    return map_tree_sitter_to_uast(
        root=root,
        language="cpp",
        map_node_kind=map_node_kind,
        file=file,
        extract_attributes=extract_type_attributes,
    )

File: crates/topos-core/src/graphs/uast/mapper_cpp.rs
//! C++ -> UAST mapper.

use std::collections::HashMap;

use tree_sitter::Node;

use super::mapper_common::map_tree_sitter_to_uast;
use super::models::{AttributeValue, UASTNode};

/// Real tree-sitter-cpp grammar node names (verified against the
/// vendored grammar directly -- the previous table here was copy-pasted
/// from the Python/Rust mappers and used node names tree-sitter-cpp
/// doesn't emit at all, e.g. `class_definition`/`struct_item`, so every
/// C++ type declaration silently fell to `Unknown`; see issue #158).
fn declaration_kind(kind: &str) -> Option<&'static str> {
    match kind {
        // Free functions AND in-class method *definitions* (with a
        // body) -- tree-sitter-cpp doesn't distinguish the two at this
        // node.
        "function_definition" => Some("FunctionDecl"),
        "class_specifier" => Some("TypeDecl"),
        "struct_specifier" => Some("TypeDecl"),
        "enum_specifier" => Some("TypeDecl"),
        "union_specifier" => Some("TypeDecl"),
        _ => None,
    }
}

/// `declaration` (free-standing) and `field_declaration` (inside a
/// class body) are dual-purpose in the grammar: a var/member
/// declaration (`int x;`) and a declaration-only fn/method signature
/// with no body (`void f(int);`, or a pure-virtual
/// `virtual double area() const = 0;`) both use the same wrapper node,
/// distinguished only by whether a `function_declarator` child is
/// present.
fn is_declaration_only_kind(kind: &str) -> bool {
    matches!(kind, "declaration" | "field_declaration")
}

/// True for real fn/method declarations, false for fn pointers.
///
/// tree-sitter-cpp represents both `void f(int);` and
/// `void (*handler)(int);` with a `function_declarator` wrapper. A real
/// prototype has the fn name directly under that wrapper; a fn pointer
/// nests the identifier under `pointer_declarator` /
/// `parenthesized_declarator` and should remain a `VarDecl`.
fn is_named_function_declarator(node: &Node) -> bool {
    if node.kind() != "function_declarator" {
        return false;
    }
    let mut cursor = node.walk();
    let found = node
        .children(&mut cursor)
        .any(|child| matches!(child.kind(), "identifier" | "field_identifier"));
    found
}

fn has_function_declarator(node: &Node) -> bool {
    let mut cursor = node.walk();
    let found = node
        .children(&mut cursor)
        .any(|child| is_named_function_declarator(&child));
    found
}

const STATEMENT_TYPES: &[(&str, &str)] = &[
    ("if_statement", "IfStmt"),
    ("for_statement", "ForStmt"),
    ("while_statement", "WhileStmt"),
    ("return_statement", "ReturnStmt"),
    ("break_statement", "BreakStmt"),
    ("continue_statement", "ContinueStmt"),
    ("throw_statement", "ThrowStmt"),
    ("try_statement", "TryStmt"),
    ("expression_statement", "ExprStmt"),
];

const EXPRESSION_TYPES: &[(&str, &str)] = &[
    ("assignment_expression", "AssignExpr"),
    ("binary_expression", "BinaryExpr"),
    ("unary_expression", "UnaryExpr"),
    ("call_expression", "CallExpr"),
    ("field_expression", "MemberExpr"),
    ("subscript_expression", "MemberExpr"),
];

pub fn map_node_kind(kind: &str) -> &'static str {
    if let Some(mapped) = declaration_kind(kind) {
        return mapped;
    }
    if is_declaration_only_kind(kind) {
        // ponytail: `map_tree_sitter_to_uast`'s `map_node_kind` callback
        // only sees the kind string, not the `Node` itself, so this
        // can't inspect children to distinguish a declaration-only
        // function prototype (`void f(int);`) from a plain var/member
        // declaration the way Python's `_has_function_declarator` check
        // does -- always mapping to VarDecl undercounts FunctionDecl for
        // header-only prototypes and pure-virtual method signatures.
        // Upgrade path: widen `map_node_kind`'s signature to take
        // `&Node` across every mapper if this undercounting turns out to
        // matter for a real corpus (abstractness detection itself does
        // not depend on this -- see `has_pure_virtual_method` below,
        // which inspects the raw `Node` directly).
        return "VarDecl";
    }
    if let Some((_, mapped)) = STATEMENT_TYPES.iter().find(|(k, _)| *k == kind) {
        return mapped;
    }
    if let Some((_, mapped)) = EXPRESSION_TYPES.iter().find(|(k, _)| *k == kind) {
        return mapped;
    }
    match kind {
        "identifier" => "Identifier",
        "translation_unit" => "File",
        s if s.ends_with("literal") || matches!(s, "string" | "integer" | "float") => "Literal",
        _ => "Unknown",
    }
}

/// Martin Abstractness classification (issue #124/#158): a class/struct
/// is abstract iff it declares at least one pure-virtual method
/// (`virtual ... = 0;`) -- the C++ idiom for an interface/abstract base
/// class. `enum`/`union` are always concrete.
fn type_kind(kind: &str) -> Option<&'static str> {
    match kind {
        "class_specifier" => Some("class"),
        "struct_specifier" => Some("struct"),
        "enum_specifier" => Some("enum"),
        "union_specifier" => Some("union"),
        _ => None,
    }
}

/// True for a declaration-only method signature with a `= 0`
/// pure-specifier, e.g. `virtual double area() const = 0;`.
fn is_pure_virtual(field_decl: &Node, source: &[u8]) -> bool {
    if !has_function_declarator(field_decl) {
        return false;
    }
    let mut cursor = field_decl.walk();
    let found = field_decl
        .named_children(&mut cursor)
        .any(|child| child.kind() == "number_literal" && child.utf8_text(source) == Ok("0"));
    found
}

fn has_pure_virtual_method(type_node: &Node, source: &[u8]) -> bool {
    let mut cursor = type_node.walk();
    for child in type_node.children(&mut cursor) {
        if child.kind() != "field_declaration_list" {
            continue;
        }
        let mut member_cursor = child.walk();
        let found = child
            .named_children(&mut member_cursor)
            .any(|member| member.kind() == "field_declaration" && is_pure_virtual(&member, source));
        if found {
            return true;
        }
    }
    false
}

fn extract_type_attributes(node: &Node, source: &[u8]) -> HashMap<String, AttributeValue> {
    let Some(kind) = type_kind(node.kind()) else {
        return HashMap::new();
    };
    let is_class_like = matches!(node.kind(), "class_specifier" | "struct_specifier");
    let resolved = if is_class_like && has_pure_virtual_method(node, source) {
        "abstractClass"
    } else {
        kind
    };
    HashMap::from([(
        "typeKind".to_string(),
        AttributeValue::Str(resolved.to_string()),
    )])
}

pub fn map_cpp_tree_to_uast(root: Node, source: &[u8], file: Option<&str>) -> UASTNode {
    map_tree_sitter_to_uast(
        root,
        "cpp",
        map_node_kind,
        source,
        file,
        None,
        Some(&extract_type_attributes),
    )
}
